Importação de NF-e cancelada

hugouchoasborges opened the pull request:
O que funciona:
- Importação apenas do XML do cancelamento da NF-e (tendo o documento já importado) pelo wizard 'Consultar Documento'
- Importação de uma pasta contendo diversos arquivos, entre eles o XML de cancelamento (submenu Importa NF-e)

O campo 'data_hora_cancelamento' ficou como pendente, faltando fazer sua conversão de String no formato ISO para fields.Datetime

hugouchoasborges committed:
[ADD] importa_nfe_cancelada

hugouchoasborges committed:
[ADD] Tratando os arquivos XML de Cancelamento de NF-e nos métodos de importação de NF-e

mileo commented on sped_nfe/models/sped_importa_nfe.py:
```diff
@@ -52,6 +58,42 @@ class ImportaNFe(models.Model):
+
+
+            documento.justificativa = nfe.evento.infEvento.detEvento.xJust
+            self.protocolo_cancelamento = nfe.retEvento.infEvento.nProt
```
Creio que isto esteja incorreto

mileo commented:
```
data_cancelamento = retevento.infEvento.dhRegEvento.valor
data_cancelamento = UTC.normalize(data_cancelamento)
self.data_hora_cancelamento = data_cancelamento            
```

hugouchoasborges committed:
[FIX] Setando o protocolo_cancelamento da nota

hugouchoasborges committed:
[FIX][ADD] Campo data_hora_cancelamento

mileo commented on sped_nfe/models/sped_importa_nfe.py:
```diff
@@ -52,6 +60,44 @@ class ImportaNFe(models.Model):
+        for documento in documentos:
+            _logger.info(u'Importando NF-e cancelada')
+
+            nome_arquivo = chave + '-01-proc-can.xml'
```
Nem sempre o número do evento vai ser o 01

mileo commented on sped_nfe/models/sped_importa_nfe.py:
```diff
@@ -70,8 +116,14 @@ def find_files(directory, pattern):
+
         for filename in find_files(self.caminho, '*.xml'):
             try:
+                if 'proc-can' in filename:
```
Nem todo xml pode ter o nome correto. 

O ideal é que vc tente processar o arquivo, então ao meu ver esta lógica toda deveria estar dento do le_nfe.

hugouchoasborges committed:
[FIX][IMP] Forçando a importação de quaisquer arquivos XML.

- Quando um arquivo de Cancelamento de XML é importado ele retornará um erro específico
- O método 'importa_caminho' selecionará os XMLs numa nova lista para tentar novamente a importação no final da importação dos outros arquivos

mileo commented on sped_nfe/models/le_nfe/inherited_sped_documento.py:
```diff
@@ -42,10 +44,55 @@
 
+        nfe = objectify.fromstring(xml)
+
+        if getattr(nfe, 'evento', None):
```
Este código não deveria mais a baixo, depois da confirmação do modelo do documento?

hugouchoasborges replied:
Acontece que se o documento for um XML de cancelamento, ele não carrega a informação do modelo do documento.

Ou seja, se o código for pra baixo da confirmação do modelo, o método sempre retornará antes de chegar no código, falando que o Modelo não é suportado antes mesmo de verificar que é um XML de cancelamento

mileo commented on sped_nfe/models/sped_importa_nfe.py:
```diff
@@ -68,27 +68,44 @@ def find_files(directory, pattern):
+                except Exception as e:
+                    if 'Nenhum documento encontrado' in (e.message or e.name):
+                        arquivos_cancelamento.append(filename)
+                    print u"Exception:  " + filename
```
Pode colocar um log.error ao invés de um print?

mileo commented on sped_nfe/models/sped_importa_nfe.py:
```diff
@@ -68,27 +68,44 @@ def find_files(directory, pattern):
-            except Exception as e:
-                print u"Exception:  " + filename
+
+        arquivos_cancelamento = []
```
Não seria melhor chamarmos de arquivos_eventos ? Pois podemos ter mais eventos alem do cancelamento

hugouchoasborges committed:
[IMP] Substituindo 'print' por '_logger.error'

hugouchoasborges committed:
[IMP] renomeando variavel: arquivos_cancelamento --> arquivos_eventos
